src/RoomMembers.ts

```diff
@@ -0,0 +1,270 @@
+import { MatrixClient } from "matrix-bot-sdk";
+
+enum Action {
+    Join,
+    Leave,
+    Other
+}
+
+const LEAVE_OR_BAN = ['leave', 'ban'];
+
+/**
+ * Storing a join event.
+ *
+ * We use `timestamp`:
+ * - to avoid maintaining tens of thousands of in-memory `Date` objects;
+ * - to ensure immutability.
+ */
+export class Join {
+    constructor(
+        public readonly userId: string,
+        public readonly timestamp: number
+    ) { }
+}
+
+/**
+ * A data structure maintaining a list of joins since the start of Mjölnir.
+ *
+ * This data structure is optimized for lookup up of recent joins.
+ */
+class RoomMembers {
+    /**
+     * The list of recent joins, ranked from oldest to most recent.
+     *
+     * Note that a user may show up in both `_joinsByTimestamp` and `_leaves`, in which case
+     * they have both joined and left recently. Compare the date of the latest
+     * leave event (in `_leaves`) to the date of the join to determine whether
+     * the user is still present.
+     *
+     * Note that a user may show up more than once in `_joinsByTimestamp` if they have
+     * left and rejoined.
+     */
+    private _joinsByTimestamp: Join[] = [];
+    private _joinsByUser: Map<string /* user id */, number /* timestamp */> = new Map();
+
+    /**
+     * The list of recent leaves.
+     *
+     * If a user rejoins and leaves again, the latest leave event overwrites
+     * the oldest.
+     */
+    private _leaves: Map<string /* user id */, number /* timestamp */> = new Map();
+
+    /**
+     * Record a join.
+     */
+    public join(userId: string, timestamp: number) {
+        this._joinsByTimestamp.push(new Join(userId, timestamp));
+        this._joinsByUser.set(userId, timestamp);
+    }
+
+    /**
+     * Record a leave.
+     */
+    public leave(userId: string, timestamp: number) {
+        if (!this._joinsByUser.has(userId)) {
+            // No need to record a leave for a user we didn't see joining.
+            return;
+        }
+        this._leaves.set(userId, timestamp);
+        this._joinsByUser.delete(userId);
+    }
+
+    /**
+     * Run a cleanup on the data structure.
+     */
+    public cleanup() {
+        if (this._leaves.size === 0) {
+            // Nothing to do.
+            return;
+        }
+        this._joinsByTimestamp = this._joinsByTimestamp.filter(join => this.isStillValid(join));
+        this._leaves = new Map();
+    }
+
+    /**
+     * Determine whether a `join` is still valid or has been superseded by a `leave`.
+     *
+     * @returns true if the `join` is still valid.
+     */
+    private isStillValid(join: Join): boolean {
```

Comment:
curiousity; how do we end up processing a join after a leave?

Reply:
what would make that "invalid"?

Reply:
In that case, we're checking whether there has been a `leave` after a `join`. If so, the `join` date isn't valid anymore.